pkg/model/runtime.go

```diff
@@ -0,0 +1,201 @@
+func InitObjects(ctx context.Context, backstageMeta bsv1alpha1.Backstage, backstageSpec *DetailedBackstageSpec, ownsRuntime bool, isOpenshift bool) (*RuntimeModel, error) {
+
+	// 3 phases of Backstage configuration:
+	// 1- load from Operator defaults, modify metadata (labels, selectors..) and namespace as needed
+	// 2- overlay some/all objects with Backstage.spec.rawRuntimeConfig CM
+	// 3- override some parameters defined in Backstage.spec.application
+	// At the end there should be an array of runtime Objects to apply (order optimized)
+
+	lg := log.FromContext(ctx)
+
+	//objectList := make([]BackstageObject, 0)
+	model := &RuntimeModel{Objects: make([]BackstageObject, 0)}
+
+	// looping through the registered runtimeConfig objects initializing the model
+	for _, conf := range runtimeConfig {
+
+		// creating the instance of backstageObject
+		backstageObject := conf.ObjectFactory.newBackstageObject()
+		var defaultErr error
+		var overlayErr error
+
+		// reading default configuration defined in the default-config/[key] file
+		// mounted from the 'default-config' configMap
+		// this is a cluster scope configuration applying to every Backstage CR by default
+		if err := utils.ReadYamlFile(utils.DefFile(conf.Key), backstageObject.Object()); err != nil {
+			defaultErr = fmt.Errorf("failed to read default value for the key %s, reason: %s", conf.Key, err)
+			//lg.V(1).Info("failed reading default config", "error", err.Error())
+		}
+
+		// reading configuration defined in BackstageCR.Spec.RawConfigContent ConfigMap
+		// if present, backstageObject's default configuration will be overriden
+		overlay, overlayExist := backstageSpec.RawConfigContent[conf.Key]
+		if overlayExist {
+			if err := utils.ReadYaml([]byte(overlay), backstageObject.Object()); err != nil {
+				overlayErr = fmt.Errorf("failed to read overlay value for the key %s, reason: %s", conf.Key, err)
+			}
+		}
+
+		// throw the error if raw configuration exists and is invalid
+		// throw the error if there is invalid or no configuration (default|raw) for Mandatory object
+		// continue if there is invalid or no configuration (default|raw) for Optional object
+		// TODO separate the case when configuration does not exist (intentionally) from invalid configuration
+		if overlayErr != nil || (!overlayExist && defaultErr != nil) {
+			if conf.need == Mandatory || (conf.need == ForLocalDatabase && *backstageSpec.Database.EnableLocalDb) {
+				return nil, errors.Join(defaultErr, overlayErr)
+			} else {
+				lg.V(1).Info("failed to read default value for optional key. Ignored \n", conf.Key, errors.Join(defaultErr, overlayErr))
+				continue
+			}
+		}
+
+		// do not add if ForLocalDatabase and LocalDb is disabled
+		if !backstageSpec.LocalDbEnabled() && conf.need == ForLocalDatabase {
+			continue
+		}
+
+		// do not add if ForOpenshift and cluster is not Openshift
+		if !isOpenshift && conf.need == ForOpenshift {
```

Comment:
I think we should also check if route is explicitly disabled by the user in the CR spec, and not add it if so.

Reply:
Ok but not here, it is for Phase 3: process Backstage.spec, getting final desired state...
For my understanding - what would be the reason of disabling it per CR?

Reply:
> Ok but not here, it is for Phase 3: process Backstage.spec, getting final desired state... 

Okay, that makes sense.

> For my understanding - what would be the reason of disabling it per CR?

When we demo'ed the automatic Route creation, Tomas mentioned that we should also provide a way to disable it in the CR (https://github.com/janus-idp/operator/issues/93). Users might not want to have a Route, if they want to access their Backstage instance differently.
As a general rule of thumb, we should be careful with things that we create automatically, and should also provide a way to disable that behavior in the CR.

Reply:
fixed as suggested for the time